grudge/models/euler.py

```diff
@@ -0,0 +1,329 @@
+"""Grudge operators modeling compressible, inviscid flows (Euler)"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+
+from abc import ABCMeta, abstractmethod
+
+from dataclasses import dataclass
+from arraycontext import (
+    thaw,
+    dataclass_array_container,
+    with_container_arithmetic
+)
+
+from meshmode.dof_array import DOFArray
+
+from grudge.discretization import DiscretizationCollection
+from grudge.dof_desc import DOFDesc, DISCR_TAG_BASE, as_dofdesc
+from grudge.models import HyperbolicOperator
+from grudge.trace_pair import TracePair
+
+import grudge.op as op
+
+
+# {{{ Array container for the Euler model
+
+@with_container_arithmetic(bcast_obj_array=False,
+                           bcast_container_types=(DOFArray, np.ndarray),
+                           matmul=True,
+                           rel_comparison=True)
+@dataclass_array_container
+@dataclass(frozen=True)
+class EulerContainer:
```

Comment:
```suggestion
class EulerField:
```
?

Reply:
Sure, that sounds much better: https://github.com/inducer/grudge/pull/196/commits/f5e0b62a67be603ff109b98a13883ba591858521